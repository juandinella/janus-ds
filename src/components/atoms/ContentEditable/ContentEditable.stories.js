import ContentEditable from './ContentEditable'

import {
  getTemplate,
} from '../../../helpers/storybook'

const Template = getTemplate(ContentEditable)

export default {
  title: 'Atoms/ContentEditable',
  component: ContentEditable,
  args: {
    children: 'Soy un text',
  },
  argTypes: {

  },
}

export const Default = Template.bind({})

