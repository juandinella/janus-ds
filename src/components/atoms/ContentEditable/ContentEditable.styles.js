import { css } from 'styled-components'

export default css`
  background-color: var(--bg-base);
  width: 100%;
  border-radius: var(--border-radius-xl);
  padding: var(--spacing-xxs) var(--spacing-sm);
  max-height: 140px;
  font-size: var(--font-size-md);
  color: var(--text-color-neutral-darker);
  min-height: 42px;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  justify-content: center;
`
