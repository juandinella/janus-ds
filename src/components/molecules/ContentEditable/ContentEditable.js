/* eslint-disable no-param-reassign */
import isPropValid from '@emotion/is-prop-valid'
import PropTypes from 'prop-types'
import React from 'react'
import styled from 'styled-components'
import styles from './ContentEditable.styles'

const shouldForwardProp = (prop) => isPropValid(prop)
const StyledContentEditable = styled.div.withConfig({ shouldForwardProp })`
  ${styles}
`

const ContentEditable = React.forwardRef(
  ({ id, onBlur, onChange, onKeyDown, placeholder, onInput }, ref) => {
    const handleBlur = () => {
      if (ref && ref.current && onBlur) {
        onBlur(ref.current.innerHTML)
      }
    }

    const scrollToBottom = () => {
      if (ref && ref.current) {
        ref.current.scrollTop = ref.current.scrollHeight
      }
    }

    const handlePaste = (e) => {
      e.preventDefault()

      const text = e.clipboardData.getData('text/plain')

      const selection = window.getSelection()
      const range = selection.getRangeAt(0)

      selection.deleteFromDocument()

      const textNode = document.createTextNode(text)
      range.insertNode(textNode)

      range.setStartAfter(textNode)
      range.setEndAfter(textNode)

      selection.removeAllRanges()
      selection.addRange(range)

      scrollToBottom()
    }

    return (
      <StyledContentEditable
        id={id}
        data-testid={id}
        ref={ref}
        contentEditable
        suppressContentEditableWarning="true"
        role="textbox"
        aria-label={placeholder}
        onBlur={handleBlur}
        onInput={onInput}
        onChange={onChange}
        onKeyDown={onKeyDown}
        onPaste={handlePaste}
        spellCheck={false}
        placeholder={placeholder}
      />
    )
  },
)

ContentEditable.displayName = 'ContentEditable'

ContentEditable.propTypes = {
  id: PropTypes.string,
  children: PropTypes.node,
  placeholder: PropTypes.string,
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
}

export default ContentEditable
