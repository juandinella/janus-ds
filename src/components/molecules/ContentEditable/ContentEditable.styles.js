import { css } from 'styled-components'

export default css`
  width: 100%;
  padding: var(--spacing-xs) var(--spacing-sm);
  max-height: 140px;
  font-size: var(--font-size-sm);
  flex-grow: 1;
  position: relative;
  font-family: inherit;
  user-select: text;
  white-space: pre-wrap;
  word-break: break-word;
  text-align: initial;
  overflow-y: auto;
  -webkit-font-smoothing: antialiased;
  -webkit-user-modify: read-write;
  overflow-wrap: break-word;
  -webkit-line-break: after-white-space;
  outline: 0 none;
  border-radius: 10px;

  &:empty:not(:focus):before {
    content: attr(placeholder);
    color: var(--color-neutral-500);
    font-size: var(--font-size-sm);
  }

  @supports (-webkit-touch-callout: none) {
    font-size: 16px;
  }

  &::-webkit-scrollbar {
    height: 4px;
    width: 4px;
    background-color: var(--color-neutral-300);
  }
  &::-webkit-scrollbar-thumb:vertical {
    background-color: var(--color-neutral-700);
    border-radius: 6px;
  }
`
