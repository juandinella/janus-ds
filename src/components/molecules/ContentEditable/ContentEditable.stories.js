import ContentEditable from './ContentEditable'

import { getTemplate } from '../../../helpers/storybook'

const Template = getTemplate(ContentEditable)

export default {
  title: 'Molecules/ContentEditable',
  component: ContentEditable,
  args: {
    placeholder: 'Ask your question here...',
  },
  argTypes: {},
}

export const Default = Template.bind({})
