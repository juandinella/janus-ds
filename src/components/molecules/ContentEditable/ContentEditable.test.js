import initStoryshots from '@storybook/addon-storyshots'
import { render, screen, cleanup, fireEvent } from '@testing-library/react'
import React from 'react'
import ContentEditable from './ContentEditable'

initStoryshots({
  storyKindRegex: /^Molecules\/ContentEditable$/,
})

const componentProps = {
  id: 'unique_id',
  placeholder: 'Ask your question here...',
  onChange: jest.fn(),
  onBlur: jest.fn(),
  onInput: jest.fn(),
}
afterEach(cleanup)
beforeEach(() => {
  componentProps.onChange.mockClear()
  componentProps.onBlur.mockClear()
  componentProps.onInput.mockClear()
})

describe('<ContentEditable />', () => {
  test('should render component correctly', () => {
    const { getByTestId } = render(<ContentEditable {...componentProps} />)
    const comp = getByTestId(componentProps.id)
    expect(comp).toBeInTheDocument()
    expect(comp).toHaveAttribute('contentEditable', 'true')
  })

  test('should display placeholder correctly', () => {
    const { getByTestId } = render(<ContentEditable {...componentProps} />)
    const comp = getByTestId(componentProps.id)

    expect(comp).toHaveAttribute('aria-label', componentProps.placeholder)
  })
})
