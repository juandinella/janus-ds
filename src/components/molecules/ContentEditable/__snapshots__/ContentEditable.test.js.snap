// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Molecules/ContentEditable Default 1`] = `
<div
  style={
    {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
      "position": "relative",
    }
  }
>
  <div
    aria-label="Ask your question here..."
    className="sc-1p2qa2e-0 ggcIrv"
    contentEditable={true}
    onBlur={[Function]}
    onPaste={[Function]}
    placeholder="Ask your question here..."
    role="textbox"
    spellCheck={false}
    suppressContentEditableWarning="true"
  />
</div>
`;
